python_modules/dagster/dagster/_core/execution/context/system.py

```diff
@@ -889,6 +891,10 @@ def step_exception(self) -> Optional[BaseException]:
     def step_output_capture(self) -> Optional[Dict[StepOutputHandle, Any]]:
         return self._step_output_capture
 
+    @property
+    def step_output_metadata_capture(self) -> Optional[Dict[StepOutputHandle, Any]]:
```

Comment:
It's possible that the `get_output_metadata` method could work instead of this. Would you be able to try that as part of your tests? If `get_output_metadata` behaves the same as what you've implemented here, then we should use `get_output_metadata` so that we aren't duplicating work

Reply:
**I looked into this, here's what I found:**

## get_output_metadata does not work for Output-attached metadata
`StepExecutionContext._output_metadata` does not get populated at op returns/yields (i.e. `yield Output(val, metadata=metadata)`), so calling `get_output_metadata` in those cases returns nothing for our hooks.

`StepExecutionContext._output_metadata`  is **only** populated in cases where `StepExecutionContext.add_output_metadata` is used to populate metadata within the op.

## What it would take to refactor this:
I see two potential refactors to solve this:
1. We could call `StepExecutionContext.add_output_metadata(metadata=output.metadata)` in `execute_step.py`, however, this raises the following error currently: 
```
dagster._core.errors.DagsterInvariantViolationError: In op 'op_name', attempted to log output metadata for output 'result' which has already been yielded. Metadata must be logged before the output is yielded.
```
This is because `StepExecutionContext.add_output_metadata` is explicitly designed to prevent out-of-op use. 

2. Alternatively, we could refactor `StepExecutionContext._output_metadata` to take StepOutputHandle keys:
```
self._output_metadata: Dict[StepOutputHandle, Any]] = {}
```
and then fit it neatly into `execute_step.py`.

In all cases, this appears to require significant refactoring of how `StepExecutionContext.add_output_metadata` works currently, which I'm not sure we want to do here.